### PR TITLE
perf(compaction): support partition in compaction group

### PR DESCRIPTION
**Pull request by soundOfDestiny**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Support partition in compaction group.
State table will be partitioned by vnode.
View https://github.com/risingwavelabs/risingwave/pull/5335 as reference.
Partition setting can either be static or be dynamic.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/meta/src/hummock/compaction/picker/base_level_compaction_picker.rs [hunk 1]
```diff
@@ -53,7 +53,11 @@ impl CompactionPicker for LevelCompactionPicker {
         }
 
         if self.config.split_by_state_table {
-            let mut member_table_ids = levels.member_table_ids.clone();
+            let mut member_table_ids = levels
+                .member_tables
+                .iter()
```
Comment (Little-Wallace) [inline src/meta/src/hummock/compaction/picker/base_level_compaction_picker.rs:58]: why not clone it directly

Comment (soundOfDestiny) [reply src/meta/src/hummock/compaction/picker/base_level_compaction_picker.rs:58]: > why not clone it directly

We can clone it directly in the future for further optimization.

File: src/storage/src/hummock/sstable/multi_builder.rs [hunk 2]
```diff
@@ -146,10 +159,42 @@ where
...
+                self.vnode_split_point = VirtualNode::MAX.to_index();
+            }
+        }
+        if self.vnode_split_point != VirtualNode::MAX.to_index() {
+            let key_vnode = full_key.user_key.get_vnode_id();
+            if key_vnode > self.vnode_split_point {
+                switch_builder = true;
+
+                let table_weight = *self.table_weights.peek().unwrap().1 as usize;
+                let (basic, remainder) = VirtualNode::COUNT.div_rem(&table_weight);
```
Comment (zwang28) [inline src/storage/src/hummock/sstable/multi_builder.rs:186]: Can table_weight be 0?

File: proto/hummock.proto [hunk 3]
```diff
@@ -61,13 +61,13 @@
   CompactionConfig group_config = 1;
   // If `parent_group_id` is not 0, it means `parent_group_id` splits into `parent_group_id` and this group, so this group is not empty initially.
   uint64 parent_group_id = 2;
-  repeated uint32 table_ids = 3;
+  repeated StateTableInfo tables = 3;
```
Comment (zwang28) [inline proto/hummock.proto:64]: Backward compatibility is broken.

File: src/storage/src/hummock/sstable/multi_builder.rs (hunk 2 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/sstable/multi_builder.rs:186]: > Can table_weight be 0?

When `table_weight==0`, it's unreachable.

File: src/storage/src/hummock/sstable/multi_builder.rs [hunk 4]
```diff
@@ -73,28 +75,42 @@ where
...
             UserKey::default()
         } else {
             FullKey::decode(&key_range.left).user_key.to_vec()
         };
 
+        if !is_target_level_l0_or_lbase {
+            split_weight_by_vnode = 0;
+        }
+
+        if split_weight_by_vnode > 0 {
```
Comment (Li0k) [inline src/storage/src/hummock/sstable/multi_builder.rs:110]: Shall we need to split sst by table_id when enable `split_weight_by_vnode` ? @Little-Wallace

File: src/meta/src/hummock/compaction/compaction_config.rs [hunk 5]
```diff
@@ -66,6 +66,7 @@ impl CompactionConfigBuilder {
                 max_sub_compaction: DEFAULT_MAX_SUB_COMPACTION,
                 max_space_reclaim_bytes: DEFAULT_MAX_SPACE_RECLAIM_BYTES,
                 split_by_state_table: false,
+                split_weight_by_vnode: 0,
```
Comment (Li0k) [inline src/meta/src/hummock/compaction/compaction_config.rs:69]: How will `split_weight_by_vnode` be calculated in this pr? I don't seem to find the relevant logic.

File: src/storage/src/hummock/sstable/multi_builder.rs [hunk 6]
```diff
@@ -162,6 +182,29 @@ where
...
             self.last_table_id = full_key.user_key.table_id.table_id;
             switch_builder = true;
+
+            if self.split_weight_by_vnode > 1 {
+                self.largest_vnode_in_current_partition =
+                    VirtualNode::COUNT / (self.split_weight_by_vnode as usize) - 1;
+            }
+        }
+        if self.largest_vnode_in_current_partition != VirtualNode::MAX.to_index() {
+            let key_vnode = full_key.user_key.get_vnode_id();
```
Comment (Li0k) [inline src/storage/src/hummock/sstable/multi_builder.rs:192]: The processing of each key will add an additional `get_vnode_id` call, we can verify whether a performance regression through the mirco bench.

We also need a benchmark to verify whether the modification is effective.

File: src/meta/src/hummock/compaction/compaction_config.rs (hunk 5 above)
Comment (soundOfDestiny) [reply src/meta/src/hummock/compaction/compaction_config.rs:69]: > How will `split_weight_by_vnode` be calculated in this pr? I don't seem to find the relevant logic.

will always be 0

File: src/storage/src/hummock/sstable/multi_builder.rs (hunk 6 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/sstable/multi_builder.rs:192]: > The processing of each key will add an additional `get_vnode_id` call, we can verify whether a performance regression through the mirco bench.
> 
> We also need a benchmark to verify whether the modification is effective.

That's why I put it in `if self.largest_vnode_in_current_partition != VirtualNode::MAX.to_index() {`
When `split_weight_by_vnode==0`, we do not need this `get_vnode_id` call.

File: src/storage/src/hummock/sstable/multi_builder.rs (hunk 4 above)
Comment (soundOfDestiny) [reply src/storage/src/hummock/sstable/multi_builder.rs:110]: > Shall we need to split sst by table_id when enable `split_weight_by_vnode` ? @Little-Wallace

It's just a virtual flag to help us make code logic more clear.

**Review by hzxa21 — COMMENTED** [review]
Generally LGTM.

Since this PR is related to perf, I think micro-bench and perf result are required before enabling this optimization. Based on the codes, turning off this optimization with `split_weight_by_vnode` set to 0 won't affect existing logics so I think it is okay to merge the PR with that setting. If this is the plan, we can create a tracking issue for the perf evaluation plan and link it to this PR so we know when it is safe to turn on the optimization.

**Review by Li0k — APPROVED** [review]
LGTM

**Pull request merged**